tests/func/test_gc.py

```diff
@@ -25,6 +25,11 @@ def setUp(self):
             self.dvc.odb.local.hash_to_path(md5)
             for md5 in self.dvc.odb.local.all()
         ]
+        for hash_info in [
+            hash_info for hash_info in self.good_cache if ".dir" in hash_info
+        ]:
+            untyped_hash_info = hash_info.replace(".dir", "")
+            self.good_cache.remove(untyped_hash_info)
```

Comment:
This double for looks very odd
```suggestion
        for hash_info in self.good_cache:
            if hash_info.isdir:
                self.good_cache.remove(hash_info.as_raw())
```

Reply:
Also, why are we even removing it ourselves? Isn't there a nice way to adjust the test

Reply:
What you propose does not work because `odb.all()` iterates over hash info values and `self.good_cache` contains paths

How about this? It hardcodes the raw_dir_hash value, but is pretty straight forward imho
```diff
   self.dvc.add(self.FOO)
         self.dvc.add(self.DATA_DIR)
+        raw_dir_hash = "bff5a787d16460f32e9f2e62b183b1cc"
         self.good_cache = [
             self.dvc.odb.local.hash_to_path(md5)
             for md5 in self.dvc.odb.local.all()
+            if md5 != raw_dir_hash
         ]
-        for hash_info in [
-            hash_info for hash_info in self.good_cache if ".dir" in hash_info
-        ]:
-            untyped_hash_info = hash_info.replace(".dir", "")
-            self.good_cache.remove(untyped_hash_info)
 
-        self.bad_cache = []
+        self.bad_cache = [self.dvc.odb.local.hash_to_path(raw_dir_hash)]
         for i in ["123", "234", "345"]:
             path = os.path.join(self.dvc.odb.local.cache_dir, i[0:2], i[2:])
             self.create(path, i)
```

Reply:
Ended up going with the following:

```diff
         self.dvc.add(self.FOO)
-        self.dvc.add(self.DATA_DIR)
+        stages = self.dvc.add(self.DATA_DIR)
+        raw_dir_hash = stages[0].outs[0].hash_info.as_raw().value
+
         self.good_cache = [
             self.dvc.odb.local.hash_to_path(md5)
             for md5 in self.dvc.odb.local.all()
+            if md5 != raw_dir_hash
         ]
 
-        self.bad_cache = []
+        self.bad_cache = [self.dvc.odb.local.hash_to_path(raw_dir_hash)]
         for i in ["123", "234", "345"]:
             path = os.path.join(self.dvc.odb.local.cache_dir, i[0:2], i[2:])
             self.create(path, i)
```